### PR TITLE
Affinity test

### LINKED ISSUE #55520: Need e2e test for service session affinity
**Is this a BUG REPORT or FEATURE REQUEST?**:
/sig network

**What happened**:
Ref https://github.com/kubernetes/kubernetes/issues/55429, session affinity stops working when ESIPP=Local. That is being fixed by https://github.com/kubernetes/kubernetes/pull/55519.

**What you expected to happen**:
We should add an e2e test to verify that session affinity is and will be continue working when ESIPP is enabled (`service.spec.externalTrafficPolicy=Local`).

### PR DESCRIPTION
**Pull request by grayluck**
**What this PR does / why we need it**:
Add e2e test for service session affinity, including all three types of services and externalTravicPolicy=Local.

**Which issue(s) this PR fixes**:
Fixes #55520

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by MrHohn** [comment]
Great!
/ok-to-test

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: Oh, didn't realize you put all these It blocks under "ESIPP [Slow]". We should move them elsewhere (to above "services" block?), otherwise they will not run in presubmit.

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: It takes sometime for the ExternalTrafficPolicy change to be propagated across nodes, which means this step maybe run before the propagation finishes.

**Review by MrHohn — COMMENTED** [review]
Thanks, I think the overall direction seems good, added couple comments.

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: 10 is probably enough? Assuming 3 backends, the possibility of getting a true negative is already lower than 0.01%...

Comment (grayluck) [reply test/e2e/framework/service_util.go]: 10 is fair enough.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: You're right. Done.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Make sense. Done.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Moved.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Redundant values removed.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Added.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: LoadBalancer tests split into two which should fix the propagation problem.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Now they share a same logic.

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: This function mixes these two cases:
- Pod is able to get response from backends but affinity doesn't hold (return false).
- Pod is unable to get response from backends (return false).

Is it worth to return (boolean, error) to help distinguishing different cases?

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: Still thinking of early return when affinity does not hold. Probably we could return when X continuous requests hit different backends?

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: I think not. I put "Affinity shouldn't hold or connection timeout." as the error message. And if it is connection timeout, there must be something really bad happened.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: It's for loadbalancer releasing. 
```
	AfterEach(func() {
		if CurrentGinkgoTestDescription().Failed {
			framework.DescribeSvc(f.Namespace.Name)
		}
		for _, lb := range serviceLBNames {
			framework.Logf("cleaning gce resource for %s", lb)
			framework.CleanupServiceGCEResources(cs, lb, framework.TestContext.CloudConfig.Region, framework.TestContext.CloudConfig.Zone)
		}
		//reset serviceLBNames
		serviceLBNames = []string{}
	})

```

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Done.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Done. Now the function will return once it receives two different responses.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: You're right.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Actually it's `StopServeHostnameService()` that needs to be removed so that LBservice tests can share all the logic.

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: Minor: Sorry I missed the point of not defer `StopServeHostnameService` instead...`StartCustomizedServeHostnameService` also creates replication controller and that needs to be cleaned up.

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: Ahhh, I just realized in kube-proxy iptables mode we redirect all pod -> exnternal VIP traffic to the Service's ClusterIP instead...So for LB service, testing affinity from a pod doesn't actually cover the case we wanted.
https://github.com/kubernetes/kubernetes/blob/f6eed81f211d929592ddb236535f306a277ff372/pkg/proxy/iptables/proxier.go#L1126-L1138

We might need to change the LB test to send request from the test VM instead (so it would be external traffic). Let's sync-up later :)

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Oh, there's RC left. 

If StopServeHostnameService releases the loadbalancer, then we are safe to use StopServeHostnameService to release the service with RC.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Thanks :laughing:

File: test/e2e/network/service.go
Comment (MrHohn) [reply test/e2e/network/service.go]: Yeah I think `StopServeHostnameService` would be enough (it deletes the service as well). The `jig.ChangeServiceType` line exists for some legacy reasons, don't think it is required anymore.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Just committed changes. Test failed locally due to network latency. Here's part of the log.

```
Mar 15 12:12:23.287: INFO: Waiting up to 20m0s for service "service" to have a LoadBalancer
Mar 15 12:12:59.379: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:04.379: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:06.380: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:06.465: INFO: Got pod response: service-gzhzm
Mar 15 12:13:08.465: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:08.549: INFO: Got pod response: service-xnrbm
Mar 15 12:13:08.662: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:08.743: INFO: Got pod response: service-gzhzm
Mar 15 12:13:10.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:11.021: INFO: Got pod response: service-xnrbm
Mar 15 12:13:12.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:17.744: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:19.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:20.857: INFO: Got pod response: service-gzhzm
Mar 15 12:13:22.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:22.827: INFO: Got pod response: service-xnrbm
Mar 15 12:13:24.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:24.826: INFO: Got pod response: service-gzhzm
Mar 15 12:13:26.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:26.833: INFO: Got pod response: service-xnrbm
Mar 15 12:13:28.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:28.826: INFO: Got pod response: service-gzhzm
Mar 15 12:13:30.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:30.827: INFO: Got pod response: service-xnrbm
Mar 15 12:13:32.744: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:37.744: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:39.746: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:44.747: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:44.748: INFO: Testing HTTP reachability of http://35.225.221.27:80
Mar 15 12:13:49.748: INFO: Got error testing for reachability of http://35.225.221.27:80: Get http://35.225.221.27:80: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Mar 15 12:13:49.748: INFO: Could not reach HTTP service through 35.225.221.27:80 after 2s: <nil>
<Test failed here>
```
It seems that it takes more than 2s from local to connect to the LB.
Please take a look and see if this is an issue. Thx.

Comment (MrHohn) [reply test/e2e/network/service.go]: Figured out what is going on. Upon the session affinity update on service, corresponding target pool on GCP will be recreated, which introduces service downtime:
https://github.com/kubernetes/kubernetes/blob/bef0838a21a38ee341eaee79ae57d05e940d9d9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L697-L722

We are hitting the 60 seconds timeout waiting for the target pool to be recreated and take effect. Suggest we use `LoadBalancerPollTimeout` for the LB case.

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: Combine this with `TestAffinityFromPod` with one more knob? Looks pretty similar.

**Comment by MrHohn** [comment]
Oddly the nodeport test failed on AWS...

**Comment by MrHohn** [comment]
> Oddly the nodeport test failed on AWS...

My theory is that AWS e2e has a more restricted firewall setup, which doesn't open all nodeport range (30000-32767) to public internet. Hence connecting to node's public IP will fail. We could pick node's internal IP instead to bypass that.

**Comment by grayluck** [comment]
AffinityConfirmCount added to 15 since a 900 sec test will have about 1/60 chance to show false positive result.

**Review by MrHohn — APPROVED** [review]
Looks good! Two minor comments.

**Comment by MrHohn** [comment]
BTW would be great to squash the fix-up commits.

**Comment by MrHohn** [comment]
/lgtm

Needs approval from @bowei :)
/assign @bowei

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: Comment needs to match function name

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: you should document tweak. why not take in a Service and merge them? What is going on in tweak()

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: godoc

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: remove this and try to have all err vars scoped as close as possible to their uses
better is to use the if scope
```
if err := foo(); err != nil {
...
}
```

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: Why do we need to set this to 0 again here?
Should probably do if hostname == "" { hostname = response } else { error }

File: test/e2e/network/service.go
Comment (bowei) [inline test/e2e/network/service.go]: `numPods, servicePort, serviceName := 3, 80, "service"`

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: The session affinity configuration takes some time to propagate. The response sequence with a interval of 2 sec normally looks like:
```
response: service-a
response: service-c
response: service-b
response: service-a
response: service-b
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
...
```
I assume the affinity holds when 15 same response observed in a row.
I am not sure why hostname != "" means an instant error.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Fixed. Doc for tweak also added.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Doc added and formated.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Fixed.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Fixed.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: If we take service as an parameter, does that means we need to check every single attributes to make sure we are not overwriting them?

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: Can we just pass the Service into the function, rather than using tweak()? I understand it is ok in this context, but I don't want to create a pattern that could potentially become quite confusing.

You can use Clone() to copy the various API objects if you need to start with a base that has some settings and change values:
```
svc := &v1.Service { ... }

copy := svc.Clone()
copy.Spec.Type = ...
```

File: test/e2e/framework/service_util.go
Comment (bowei) [reply test/e2e/framework/service_util.go]: Are you creating a new service for each test case or testing flipping affinity on/off?

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Sure I can do that. Now service will be passed as parameter and be created outside of StartServeHostnameService.
See if I did it right way.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: I am not creating new services every time since I want to make sure session affinity can be modified after service created.
Test logic added. Now the test will test SessionAffinity=disabled -> enabled -> disabled.

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: Can we do this to avoid the logic repeated:
```
type affinityTracker struct {
  hostTrace []string
}

func (at *affinityTracker) record(host string) {
  at.hostTrace = append(at.hostTrace, host)
}

func (at *affinityTracker) check(count int) {
  if len(at.hostTrace) < count {
    return false
  }
  last := at.hostTrace[len(at.hostTrace)-count:]
  host := at.hostTrace[len(at.hostTrace)-count]
  for _, h := range last {
    if h != host {
      return false
    }
  }
  return true
}
```

Comment (bowei) [reply test/e2e/framework/service_util.go]: You can then use it in both tests

Comment (grayluck) [reply test/e2e/framework/service_util.go]: Done. Take a look.

File: test/e2e/framework/service_util.go
Comment (bowei) [inline test/e2e/framework/service_util.go]: no need to export this function (capitalized)

TestAffinity => testAffinity

Also, retry to avoid using "Test" as a prefix, as this is automatically picked up by the unit testing framework as a testing.

**Comment by grayluck** [comment]
Commits squashed.
Thanks for review!

**Comment by rmmh** [comment]
/test pull-kubernetes-typecheck

It looks like this might be failing now.

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: nit: works as expected

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: nit: clusterIp -> clusterIP

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: nit: clusterIp -> clusterIP

File: test/e2e/framework/service_util.go
Comment (MrHohn) [inline test/e2e/framework/service_util.go]: It seems we have `fulfilled` here but have `fullfilled` below...

File: test/e2e/network/service.go
Comment (MrHohn) [inline test/e2e/network/service.go]: This is something we should put into util. But I guess it is fine for now :)

**Review by MrHohn — APPROVED** [review]
/lgtm

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: This makes perfect sense. Actually not only this function but also default settings and constants should go to service_util too. But that will cause some change. I will clean it up after this PR get merged.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: fuLLfilled should be a typo. Changed to fulfilled.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Fixed.

File: test/e2e/network/service.go
Comment (grayluck) [reply test/e2e/network/service.go]: Fixed.

File: test/e2e/framework/service_util.go
Comment (grayluck) [reply test/e2e/framework/service_util.go]: Fixed.

File: test/e2e/network/service.go
Comment (MrHohn) [reply test/e2e/network/service.go]: SGTM

**Pull request merged**
